### ISSUE TITLE
Return Array<T>

### ISSUE THREAD

**Issue by Lelelo1** [haxiomic/haxe-c-bridge#34]
I am noticing I get a message from setting the return type to `Array<String>` of a method, when compiling:

> Array<T> is not supported for C export, try using cpp.Pointer<T> instead

When try to walk around using a `List<T>` I can compile, but the method's return type is given as `HaxeObject`

Is there some way I can construct a suitable array to be returned - and receiving array or list out from the haxe environment?

**Comment by haxiomic** [comment]
Hey :), I've added a test to show you how to do this

```haxe
static public function getHaxeArray(length: Star<Int>) {
	var array = [1,2,3,4,5];
	Native.set(length, array.length);
	return cpp.Pointer.ofArray(array);
}
```

I've disallowed passing arrays directly because you lose the information about the length of the array, however maybe by default I should just do `cpp.Pointer.ofArray(array)` automatically and leave it up to the user to pass the array length. I was thinking about returning a struct to store both the pointer and the length

**Comment by haxiomic** [comment]
* updated my earlier answer

**Comment by Lelelo1** [comment]
Interesting, thanks for the details.

Can I use structure?

**Comment by haxiomic** [comment]
Absolutely, here's how I plan on passing arrays in the future:

- Define type in C (this works for any array type, but we might want to generate type specialized versions like `HaxeArrayInt`
```C
typedef struct HaxeArray {
	const void* ptr;
	int64 length;
} HaxeArray;
```

- When passing an array to C we:
    - Make any necessary type conversions, for haxe objects like Strings we need to convert to `const char*`
    - Get the pointer to the underlying data, using `cpp.Pointer.ofArray(array)`
    - Keep a ref to the array so the data doesn't get garbage collected, this is what `HaxeCBridge.retainHaxeObject(array);` does
    - Return a stack allocated struct of HaxeArray ([example](https://github.com/haxiomic/haxe-c-bridge/blob/631cb58ba703de097b2e062ac7e9531c93b6c5c3/test/unit/src/Main.hx#L189))

- When passing one of these HaxeArrays from C to Haxe, we:
     - Use [`cpp.NativeArray.setUnmanagedData()`](https://api.haxe.org/cpp/NativeArray.html#setUnmanagedData) to create an Array<T> from our ptr and length

**Comment by Lelelo1** [comment]
I assume this file is needed to define the return type: https://github.com/haxiomic/haxe-c-bridge/blob/631cb58ba703de097b2e062ac7e9531c93b6c5c3/test/unit/src/MessagePayload.h#L3 ?

.. or I get `HaxeObject` as return type in the generated header?

**Comment by haxiomic** [comment]
Yeah, the MessagePayload is an example of working a custom C struct in haxe, checkout how it's used in app.c and Main.hx

You'd be doing the same thing but with a new struct called HaxeArray

**Comment by Lelelo1** [comment]
Is there a possibility one can provide a type declaration in the HaxeArray struct:

```
typedef struct HaxeArray {
	const void* ptr;
	const Employee; // <---
	int64 length;
} HaxeArray;
```



... Employee struct:

```
typedef struct {
    char name[NAMESIZE];
    char sex;
} Person;

typedef struct {
    Person person;
    char job[JOBSIZE];
} Employee;
```

Can I then make a cast of `ptr` and initialize it to the given `length` when consuming the output?

```
HaxeArray fromHaxe; // from

Employee employees[fromHaxe.length] = (some cast to array of Employee) fromHaxe.prt;

// use employees with type declaration

```

**Comment by haxiomic** [comment]
You can't store a type reference in C like that (it'll complain at the syntax level) but you could create an enum for this, like

```C
enum ArrayType {
    EMPLOYEE,
    PERSON,
};

typedef struct HaxeArray {
	const void* ptr;
	const ArrayType arrayType; // <---
	int length;
} HaxeArray;

Then we can know what array type we have if we switch on the arrayType field
```

**Comment by Lelelo1** [comment]
```
// above struct definition of employee

typedef struct HaxeArray {
	const void* ptr;
	Employee type; // <--- (corrected)
	int64 length;
} HaxeArray;

```

I read about both anonymous structs and nested typedef structs as well, in C.

From I own current use case - to pass an array of models out of haxe, I view this as a serialization area. It reminds be of web request, json, and react-native bridge. Meaning there is highlevel in Haxe, and low level C - and in the end you want higher programming features and data, after consuming C.

The `type` would be just a dummy variable not storing anything, but allowing casting of `prt` to it, when consuming C

**Comment by Lelelo1** [comment]
```
typedef struct MessagePayload {
	float someFloat;
	char cStr[10];
} MessagePayload; // everything ok


typedef struct StructWithTypeInside { // variables that gets declared with the type becomes void pointers in the generated heading file?
	MessagePayload messagePayload; // -----> ? messagePayload; // Or is the problem here - that type is lost?
} StructWithTypeInside 
```

**Comment by haxiomic** [comment]
The problem only is when you have some data passed to C where you’ve lost the type. So you don’t have to lose the type at all, you can have a function in haxe like getEmployees() and this could return a struct that contains a pointer to a set of Employees. So in your struct, const void* Ptr would become const Employee* ptr.

However, if you want to have something like getStuff(): Array<Dynamic>, where you don’t explicitly say what the type is when you pass to C, then you’ve lose type information and you need to store it somewhere

**Comment by Lelelo1** [comment]
I think I might have understood partially, looking at  [MessagePayload](https://github.com/haxiomic/haxe-c-bridge/tree/main/test/unit/src) again.

Then in [haxe code](https://github.com/haxiomic/haxe-c-bridge/blob/main/test/unit/src/Main.hx#L171) it is exposed as a lambda function .

```
import cpp.Callable;
// ...
fnStruct: Callable<MessagePayload -> Void>
```

And in [app.c](https://github.com/haxiomic/haxe-c-bridge/blob/main/test/unit/app.c), client

![Screen Shot 2021-10-09 at 13 00 10](https://user-images.githubusercontent.com/42244983/136656520-a951c6c6-79f1-4a8c-9b95-9cfb893c7246.png)

... the type stays

returning just `messagePayload` would on the other hand cause the type to be lost.

---

If I understood correctly I can declare a pointer inside `MessagePayload` and it can contain type of int, float etc, but also a struct? So could have list of User struct, declared somewhere and a pointer to i  MessagePayload

```
fnStruct: Callable<MessagePayload -> Void> 
```

But you are saying also I can add a pointer to any array, of int, float or struct - `inside MessageLoader`..?

**Comment by haxiomic** [comment]
> When it comes to keeping it in memory, can you go around the problem - by copying the array in the client/consuming application?

Absolutely, for example, you may stack allocate an array in haxe and return it, that way it'll be valid so long as the variable that references it in C stays in scope
https://en.wikipedia.org/wiki/Variable-length_array#:~:text=Implementation%5Bedit%5D-,C99,-%5Bedit%5D

Then you can copy it / do whatever you wish without worrying about allocation and haxe

**Comment by Lelelo1** [comment]
Do you heave any new ideas on how to handle this?
What I am looking for is to have shared logic (in Haxe) that can use used from swift, with typings. Is it simply a limitation in C language that makes this tricky?

**Comment by haxiomic** [comment]
Aye no way around this, it's simply C limitations making this tricky, so you need to use one of the solutions talked about here: passing the array length along with the array pointer

**Comment by singpolyma** [comment]
If I return  `cpp.Pointer.ofArray(someArray)` from a function, it seems that haxe-c-bridge does not call retain on this automatically. I guess it just assumes if I am a Pointer that all bets are off?  It seems like Pointer (vs RawPointer) is meant to be to something on the GC head and so should be auto retained still, but maybe I'm wrong?
